src/TokenManagers/XboxTokenManager.js

```diff
@@ -68,32 +68,43 @@ class XboxTokenManager {
   }
 
   async verifyTokens (relyingParty) {
+    if (this.forceRefresh) return false
     const ut = await this.getCachedUserToken()
     const xt = await this.getCachedXstsToken(relyingParty)
-    if (!ut || !xt || this.forceRefresh) {
-      return false
-    }
     debug('[xbl] have user, xsts', ut, xt)
-    if (ut.valid && xt.valid) {
+    if ((ut.valid && xt.valid) || xt.valid) return true
```

Comment:
Before if the user token was invalid but the XSTSToken was ok it would fail the check and refresh both tokens now we only refresh both tokens once the XSTSToken is invalid. This logic change also doesn't force a user token refresh if an XSTS token doesn't exist, for example, say we had an XSTSToken generated with one relyingParty and then wanted to create another one with a different RP previously the `getCachedXstsToken` would've returned undefined meaning the check would fail thus causing the creation of a new user token and an XSTS token. Actually I guess all we need here is `if (xt.valid)`

Reply:
Sounds to me like intended behavior then. `verifyToken(s)` is supposed to check that the state of all the tokens it's managing are valid, and refresh them as needed. If this returns false, the code walks back an authentication step to correct it. Calls to neither `getUserToken()` or `getXSTSToken` are guaranteed to work until `verifyTokens()` is called. In addition if the user token is invalid, this usually the means the XSTS token will also be expiring soon and should be refreshed.

Reply:
If the validity duration of the user/xsts token has a big disreprency though (so the XSTS token lasts alot longer than the user token) maybe it would make sense to add a separate verify method just for the XSTS token. Otherwise with this change, getXSTSToken() called with cached user token would fail to work even if verifyTokens() is called.

Reply:
I don't see how `verifyTokens` guarantees that `getUserToken` will work as it doesn't validate msa tokens which is what is used to gen a user token.

I don't understand why we refresh tokens when we don't need to. XSTS tokens are valid for 16 hours whereas user, title and device tokens are valid for 14 days. Ideally, we should be checking if the XSTS token exists, if not we then verify the "constructor" tokens and then create a new XSTS token.

We only need to refresh the user token when the XSTS token is invalid because the UT isn't needed until that point.

Reply:
`verifyTokens` here more specifically means "vertifyTokensInCache". If you call this API method and use these cached tokens, the implication is they are valid. `getUserToken` or `getXSTSToken` take their prerequisite tokens as parameters, designed to be obtained from the cache. It might mean one less API call for the purpose of the authentication flow, but this is a breaking change from how the API is designed. If the user token outlives the XSTS token, I don't see the problem here.